File: internal/ot/transform.go
package ot

import "fmt"

// Transform resolves two operations made concurrently against the same
// document, returning aPrime and bPrime such that:
//
//	Compose(a, bPrime) ≡ Compose(b, aPrime)
//
// which is the TP1 property, and the whole reason this package exists: it lets
// two peers who edited the same text at the same time end up with byte-identical
// documents regardless of the order the edits arrived in.
//
// Note the asymmetry in the insert/insert case below. When both sides insert at
// the same position there is no correct answer, only a consistent one — so a's
// insert is always placed first. Every peer must apply that same rule; the
// TypeScript engine in web/src/ot does, and testdata/ot-vectors.json is what
// keeps the two honest.
func Transform(a, b *OpSeq) (*OpSeq, *OpSeq, error) {
	if a.baseLen != b.baseLen {
		return nil, nil, fmt.Errorf("%w: cannot transform, base lengths %d and %d differ",
			ErrLengthMismatch, a.baseLen, b.baseLen)
	}

	aPrime, bPrime := New(), New()
	ia, ib := newIter(a.ops), newIter(b.ops)

	for ia.ok || ib.ok {
		// Inserts from a take precedence, deterministically.
		if ia.ok && ia.cur.Kind == KindInsert {
			aPrime.Insert(ia.cur.S)
			bPrime.Retain(len(ia.cur.S))
			ia.next()
			continue
		}
		if ib.ok && ib.cur.Kind == KindInsert {
			aPrime.Retain(len(ib.cur.S))
			bPrime.Insert(ib.cur.S)
			ib.next()
			continue
		}
		if !ia.ok || !ib.ok {
			return nil, nil, fmt.Errorf("%w: ran out of components while transforming", ErrIncompatible)
		}

		switch {
		case ia.cur.Kind == KindRetain && ib.cur.Kind == KindRetain:
			n := min(ia.cur.N, ib.cur.N)
			aPrime.Retain(n)
			bPrime.Retain(n)
			ia.takeN(n)
			ib.takeN(n)

		case ia.cur.Kind == KindDelete && ib.cur.Kind == KindDelete:
			// Both peers deleted the same bytes. The text is already gone, so
			// neither transformed operation should delete it a second time.
			n := min(ia.cur.N, ib.cur.N)
			ia.takeN(n)
			ib.takeN(n)

		case ia.cur.Kind == KindDelete && ib.cur.Kind == KindRetain:
			n := min(ia.cur.N, ib.cur.N)
			aPrime.Delete(n)
			ia.takeN(n)
			ib.takeN(n)

		case ia.cur.Kind == KindRetain && ib.cur.Kind == KindDelete:
			n := min(ia.cur.N, ib.cur.N)
			bPrime.Delete(n)
			ia.takeN(n)
			ib.takeN(n)

		default:
			return nil, nil, fmt.Errorf("%w: transforming %s against %s", ErrIncompatible, ia.cur.Kind, ib.cur.Kind)
		}
	}

	return aPrime, bPrime, nil
}

// Transform is the method form. The receiver is the "a" side, so the first
// result is the receiver rebased onto other.
func (s *OpSeq) Transform(other *OpSeq) (*OpSeq, *OpSeq, error) { return Transform(s, other) }

File: web/src/ot/transform.ts
import { Iter, OpSeq, OTError, opLength } from "./operation";

/**
 * Resolve two operations made concurrently against the same document.
 *
 * Returns [aPrime, bPrime] such that compose(a, bPrime) ≡ compose(b, aPrime) —
 * the TP1 property. It is what lets two people type in the same paragraph and
 * still end up with byte-identical documents.
 *
 * Note the asymmetry in the insert/insert case: when both sides insert at the
 * same position there is no correct answer, only a consistent one, so a's
 * insert always goes first. The Go engine applies the same rule, and
 * testdata/ot-vectors.json is what proves the two still agree.
 */
export function transform(a: OpSeq, b: OpSeq): [OpSeq, OpSeq] {
  if (a.baseLength !== b.baseLength) {
    throw new OTError(
      `cannot transform: base lengths ${a.baseLength} and ${b.baseLength} differ`,
    );
  }

  const aPrime = new OpSeq();
  const bPrime = new OpSeq();
  const ia = new Iter(a.ops);
  const ib = new Iter(b.ops);

  while (ia.ok || ib.ok) {
    if (ia.cur?.kind === "insert") {
      aPrime.insert(ia.cur.s);
      bPrime.retain(opLength(ia.cur));
      ia.advance();
      continue;
    }
    if (ib.cur?.kind === "insert") {
      aPrime.retain(opLength(ib.cur));
      bPrime.insert(ib.cur.s);
      ib.advance();
      continue;
    }
    if (!ia.cur || !ib.cur) {
      throw new OTError("ran out of components while transforming");
    }

    const av = ia.cur;
    const bv = ib.cur;

    if (av.kind === "retain" && bv.kind === "retain") {
      const n = Math.min(av.n, bv.n);
      aPrime.retain(n);
      bPrime.retain(n);
      ia.takeN(n);
      ib.takeN(n);
    } else if (av.kind === "delete" && bv.kind === "delete") {
      // Both peers deleted the same bytes; the text is already gone, so
      // neither transformed operation should delete it a second time.
      const n = Math.min(av.n, bv.n);
      ia.takeN(n);
      ib.takeN(n);
    } else if (av.kind === "delete" && bv.kind === "retain") {
      const n = Math.min(av.n, bv.n);
      aPrime.delete(n);
      ia.takeN(n);
      ib.takeN(n);
    } else if (av.kind === "retain" && bv.kind === "delete") {
      const n = Math.min(av.n, bv.n);
      bPrime.delete(n);
      ia.takeN(n);
      ib.takeN(n);
    } else {
      throw new OTError(`cannot transform ${av.kind} against ${bv.kind}`);
    }
  }

  return [aPrime, bPrime];
}
